src/webgpu/shader/execution/shader_io/shared_structs.spec.ts

```diff
@@ -0,0 +1,361 @@
+export const description = `Test the shared use of structures containing entry point IO attributes`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { GPUTest } from '../../../gpu_test.js';
+import { checkElementsEqual } from '../../../util/check_contents.js';
+
+export const g = makeTestGroup(GPUTest);
+
+g.test('shared_with_buffer')
+  .desc(
+    `Test sharing an entry point IO struct with a buffer.
+
+     This test defines a structure that contains both builtin attributes and layout attributes,
+     and uses that structure as both an entry point input and the store type of a storage buffer.
+     The builtin attributes should be ignored when used for the storage buffer, and the layout
+     attributes should be ignored when used as an entry point IO parameter.
+    `
+  )
+  .fn(async t => {
+    // Set the dispatch parameters such that we get some interesting (non-zero) built-in variables.
+    const wgsize = new Uint32Array([8, 4, 2]);
+    const numGroups = new Uint32Array([4, 2, 8]);
+
+    // Pick a single invocation to copy the input structure to the output buffer.
+    const targetLocalIndex = 13;
+    const targetGroup = new Uint32Array([2, 1, 5]);
+
+    // The test shader defines a structure that contains members decorated with built-in variable
+    // attributes, and also layout attributes for the storage buffer.
+    const wgsl = `
+      [[block]]
+      struct S {
+        /* byte offset:  0 */ [[size(32)]]  [[builtin(workgroup_id)]] group_id : vec3<u32>;
+        /* byte offset: 32 */               [[builtin(local_invocation_index)]] local_index : u32;
+        /* byte offset: 64 */ [[align(64)]] [[builtin(num_workgroups)]] numGroups : vec3<u32>;
+      };
+
+      [[group(0), binding(0)]]
+      var<storage, read_write> outputs : S;
+
+      [[stage(compute), workgroup_size(${wgsize[0]}, ${wgsize[1]}, ${wgsize[2]})]]
+      fn main(inputs : S) {
+        if (inputs.group_id.x == ${targetGroup[0]}u &&
+            inputs.group_id.y == ${targetGroup[1]}u &&
+            inputs.group_id.z == ${targetGroup[2]}u &&
+            inputs.local_index == ${targetLocalIndex}u) {
+          outputs = inputs;
+        }
+      }
+    `;
+
+    const pipeline = t.device.createComputePipeline({
+      compute: {
+        module: t.device.createShaderModule({
+          code: wgsl,
+        }),
+        entryPoint: 'main',
+      },
+    });
+
+    // Allocate a buffer to hold the output structure.
+    const bufferNumElements = 32;
+    const outputBuffer = t.device.createBuffer({
+      size: bufferNumElements * Uint32Array.BYTES_PER_ELEMENT,
+      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC,
+    });
+    const bindGroup = t.device.createBindGroup({
+      layout: pipeline.getBindGroupLayout(0),
+      entries: [
+        {
+          binding: 0,
+          resource: {
+            buffer: outputBuffer,
+          },
+        },
```

Comment:
uNit: we typically put these structs on one line because it reads better if you have multiple bindings, though obviously you don't here.
{ binding: 0, ... },
{ binding: 1, ... },